compiler/rustc_session/src/options.rs

```diff
@@ -58,10 +58,58 @@ macro_rules! hash_substruct {
     };
 }
 
+macro_rules! gather_tmods {
+    ($_opt_name:ident, $init:expr, $opt_expr:expr, $mods:expr, [SUBSTRUCT], [TMOD]) => {
+        compile_error!("SUBSTRUCT can't be TMOD (target modifier)");
+    };
+    ($opt_name:ident, $init:expr, $opt_expr:expr, $mods:expr, [UNTRACKED], [TMOD]) => {
+        if *$opt_expr != $init {
+            $mods.push(format!("{}={:?}", stringify!($opt_name), $opt_expr));
+        }
+    };
+    ($opt_name:ident, $init:expr, $opt_expr:expr, $mods:expr, [TRACKED], [TMOD]) => {
+        if *$opt_expr != $init {
+            $mods.push(format!("{}={:?}", stringify!($opt_name), $opt_expr));
+        }
+    };
+    ($opt_name:ident, $init:expr, $opt_expr:expr, $mods:expr, [TRACKED_NO_CRATE_HASH], [TMOD]) => {
+        if *$opt_expr != $init {
+            $mods.push(format!("{}={:?}", stringify!($opt_name), $opt_expr));
+        }
+    };
+    ($_opt_name:ident, $init:expr, $opt_expr:expr, $mods:expr, [SUBSTRUCT], []) => {
+        $opt_expr.gather_target_modifiers($mods);
+    };
+    ($_opt_name:ident, $init:expr, $_opt_expr:expr, $_mods:expr, [UNTRACKED], []) => {{}};
+    ($_opt_name:ident, $init:expr, $_opt_expr:expr, $_mods:expr, [TRACKED], []) => {{}};
+    ($_opt_name:ident, $init:expr, $_opt_expr:expr, $_mods:expr, [TRACKED_NO_CRATE_HASH], []) => {{}};
+}
+
+macro_rules! gather_tmods_top_level {
+    ($_opt_name:ident, $opt_expr:expr, $mods:expr, [SUBSTRUCT], [TMOD]) => {
+        compile_error!("SUBSTRUCT can't be TMOD (target modifier)");
+    };
+    ($opt_name:ident, $opt_expr:expr, $mods:expr, [UNTRACKED], [TMOD]) => {
+        compile_error!("Top level option can't be TMOD (target modifier)");
+    };
+    ($opt_name:ident, $opt_expr:expr, $mods:expr, [TRACKED], [TMOD]) => {
+        compile_error!("Top level option can't be TMOD (target modifier)");
+    };
+    ($opt_name:ident, $opt_expr:expr, $mods:expr, [TRACKED_NO_CRATE_HASH], [TMOD]) => {
+        compile_error!("Top level option can't be TMOD (target modifier)");
+    };
+    ($_opt_name:ident, $opt_expr:expr, $mods:expr, [SUBSTRUCT], []) => {
+        $opt_expr.gather_target_modifiers($mods);
+    };
+    ($_opt_name:ident, $_opt_expr:expr, $_mods:expr, [UNTRACKED], []) => {{}};
+    ($_opt_name:ident, $_opt_expr:expr, $_mods:expr, [TRACKED], []) => {{}};
+    ($_opt_name:ident, $_opt_expr:expr, $_mods:expr, [TRACKED_NO_CRATE_HASH], []) => {{}};
+}
+
 macro_rules! top_level_options {
     ( $( #[$top_level_attr:meta] )* pub struct Options { $(
         $( #[$attr:meta] )*
-        $opt:ident : $t:ty [$dep_tracking_marker:ident],
+        $opt:ident : $t:ty [$dep_tracking_marker:ident $( $tmod:ident )*],
```

Comment:
Is the `$( )*` just because it can be optional? Can you use `?` instead?

Reply:
Yes, fixed.